assets/app/scripts/services/labelFilter.js

```diff
@@ -0,0 +1,337 @@
+      });
+    }
+  };
+
+  LabelFilter.prototype.setLabelSuggestions = function(suggestions) {
+    this._existingLabels = suggestions;
+  };
+
+  LabelFilter.prototype._extractLabelsFromItem = function(item, map) {
+    // TODO temporary until we handle everything with item.metadata.labels
+    var labels = item.labels;
+    if (item.metadata) {
+      labels = item.metadata.labels;
+    }
+
+    var self = this;
+    $each(labels, function(key, value) {
+      if (!map[key]) {
+        map[key] = [];
+      }
+      map[key].push({value: value});
+    });
+  };
+
+  LabelFilter.prototype.onActiveFiltersChanged = function(callback) {
+    this._onActiveFiltersChangedCallbacks.add(callback);
+  };
+
+  LabelFilter.prototype.filterResources = function(resources) {
+    var filteredResources = {};
+    var self = this;
+    $each(resources, function(resId, resource) {
+      if (self.isResourceIncludedInActiveFilters(resource)) {
+        filteredResources[resId] = resource;
+      }
+    });
+    return filteredResources;
+  };
+
+  LabelFilter.prototype.isResourceIncludedInActiveFilters = function(resource) {
+    if (!resource) {
+      return false;
+    }
+    var labels = resource.labels || {};
+    if (resource.metadata) {
+      labels = resource.metadata.labels || {};
+    }
+    for (var id in this._activeFilters) {
+      var filter = this._activeFilters[id];
+      switch(filter.operator) {
+        case "EXISTS":
+          if (!labels[filter.key]) {
+            return false;
+          }
+          break;
+        case "IN":
+          var keep = false;
+          if (labels[filter.key]) {
+            for (var i = 0; i < filter.values.length; i++) {
+              if (labels[filter.key] == filter.values[i]) {
+                keep = true;
+              }
+            }
+          }
+          if (!keep) {
+            return false;
+          }
+          break;
+        case "NOT_IN":
+          var keep = true;
+          if (labels[filter.key]) {
+            for (var i = 0; i < filter.values.length; i++) {
+              if (labels[filter.key] == filter.values[i]) {
+                keep = false;
+              }
+            }
+          }
+          if (!keep) {
+            return false;
+          }
+      }
+    }
+    return true;
+  };
+
+  LabelFilter.prototype.setupFilterWidget = function() {
+    var self = this;
+
+    $('.label-filter-key').selectize({
+      valueField: "key",
+      labelField: "key",
+      searchField: ["key"],
+      create: true,
+      persist: true, // i want this to be false but there appears to be a bug in selectize where setting
+                     // this to false has a side effect of causing items that were not created by the user
+                     // to also disappear from the list after being removed
+      preload: true,
+      onItemAdd: function(value, $item) {
+        var selectizeValues = $('.label-filter-values')[0].selectize;
```

Comment:
is there a way to get this from the context (`this`, etc). class queries will break if we ever have more than one in a page (even if they're hidden)